### ISSUE TITLE
[Bug Report]: Publish module to private bicep registry Job fails

### ISSUE THREAD

**Issue by msc365admin** [Azure/ResourceModules#3378]
### Describe the bug

The `[Publish module to private bicep registry]` job fails because the `Find-TemplateFile` function in the `Get-ModulesToPublish` cmdlet always collects the `main.json` file. 

```powershell
#Prioritizing the bicep file
$TemplateFilePath = Join-Path $FolderPath 'main.bicep'
if (-not (Test-Path $TemplateFilePath)) {
    $TemplateFilePath = Join-Path $FolderPath 'main.json'
}
```
Then the `Publish-ModuleToPrivateBicepRegistry` cmdlet throws an error.

```powershell
if ((Split-Path $TemplateFilePath -Extension) -ne '.bicep') {
    throw "The template in path [$TemplateFilePath] is no bicep template."
}
```
Which results in a failed job on the pipeline.

![image](https://github.com/Azure/ResourceModules/assets/120459644/450054c3-26b8-4572-83dd-93d577b2d549)

All module versions are deployed, while the `main.json` file is the last item in `$modulesToPublish`.

### To reproduce

1. Cloned the latest version on 06/27/2023-12:31 UTC+01:00:00.
2. Merged with `main` branch to get full deployment.
3. Ran into job errors for every updated module.

![image](https://github.com/Azure/ResourceModules/assets/120459644/cf4684d7-d653-47a2-b3b1-d9368e34c2c3)

### Code snippet

```Bicep
See description.
```

### Relevant log output

```Shell
2023-06-27T13:35:38.2013889Z [0m
2023-06-27T13:35:38.5600988Z [33;1mVERBOSE: Performing the operation "Publish" on target "Private bicep registry entry [bicep/modules/network.public-ip-addresses] version [latest] to registry [***]".[0m
2023-06-27T13:35:43.1768514Z The module "br:***.azurecr.io/bicep/modules/network.public-ip-addresses:latest" already exists in registry. Use --force to overwrite the existing module.
2023-06-27T13:35:43.1960229Z [33;1mVERBOSE: Publish complete[0m
2023-06-27T13:35:43.1963218Z ##[endgroup]
2023-06-27T13:35:43.1963838Z ##[group] - [network/public-ip-addresses] [0.4.0]
2023-06-27T13:35:43.1966173Z [33;1mVERBOSE: Invoke Publish-ModuleToPrivateBicepRegistry with[0m
2023-06-27T13:35:43.1978365Z [33;1mVERBOSE: {
2023-06-27T13:35:43.1978930Z   "BicepRegistryRgName": "***-sbx-rg",
2023-06-27T13:35:43.1979505Z   "BicepRegistryName": "***",
2023-06-27T13:35:43.1982028Z   "BicepRegistryRgLocation": "West Europe",
2023-06-27T13:35:43.1982615Z   "TemplateFilePath": "/agent/_work/1/s/modules/network/public-ip-addresses/main.json",
2023-06-27T13:35:43.1983139Z   "ModuleVersion": "0.4.0"
2023-06-27T13:35:43.1983465Z }
2023-06-27T13:35:43.1983763Z [0m
2023-06-27T13:35:43.2693459Z [31;1mThe template in path [/agent/_work/1/s/modules/network/public-ip-addresses/main.json] is no bicep template.[0m
2023-06-27T13:35:43.2694572Z [31;1mAt /agent/_work/1/s/utilities/pipelines/resourcePublish/Publish-ModuleToPrivateBicepRegistry.ps1:65 char:13[0m
2023-06-27T13:35:43.2695403Z [31;1m+             throw "The template in path [$TemplateFilePath] is no bic …[0m
2023-06-27T13:35:43.2696056Z [31;1m+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
2023-06-27T13:35:43.2696882Z [31;1m+ CategoryInfo          : OperationStopped: (The template in pat… no bicep template.:String) [], RuntimeException[0m
2023-06-27T13:35:43.2697820Z [31;1m+ FullyQualifiedErrorId : The template in path [/agent/_work/1/s/modules/network/public-ip-addresses/main.json] is no bicep template.[0m
2023-06-27T13:35:43.3937326Z ##[error]Script failed with exit code: 1
2023-06-27T13:35:43.3956751Z [command]/usr/bin/az account clear
2023-06-27T13:35:43.9693279Z ##[section]Finishing: Publish module to private bicep registry
```

**Comment by AlexanderSehr** [comment]
That's not great. I just checked some pipelines and weirdly enough, I cannot see the behavior on our end (for example [public-ip-addresses](https://github.com/Azure/ResourceModules/actions/runs/5370736788/jobs/9743102554#step:4:564)).

Regardless - if it works differently for different people we should anyways investigate how we can improve the robustness of the code. Most notably if we need to change something about the logic that finds the template to publish and should differentatiate the cases:
- If I find both a `main.json` & `main.bicep` file, publish the `main.bicep` file
- If I find only a `main.json` file, publish the `main.json` file (but not to the bicep registry)

**Comment by msc365admin** [comment]
Somehow, it tries to publish twice as well.
> btw: this happens in all **Azure DevOps Pipelines**, not just `public-ip-addresses` and not tested by myself with GitHub Actions.

![image](https://github.com/Azure/ResourceModules/assets/120459644/0286e2f0-11dc-4765-8dde-6dad844b75f6)

As mentioned all modules get published into the private bicep registry. 
So, no blocking issues using the latest modules 😄 

Thanks for your reply and all the hard work!

**Comment by msc365admin** [comment]
Forgot to mention that I also applied the fix of this [PR](https://github.com/Azure/ResourceModules/pull/3388/files).

**Comment by krbar** [comment]
It looks like related (or even duplicate) to #3498, which should be resolved after #3554 has been merged. Can we close this issue?

**Comment by AlexanderSehr** [comment]
I believe so. @msc365admin, both the change to check for `404` as well as a more refined handling of having both the `main.bicep` & `main.json` was merged. Thanks to @krbar.

You are right that there is no point in pushing both (especially given that 'technically speaking' all that is pushed is the compiled Bicep template anyways and hence both templates are the same). Anyhow - the logic now works as follows
- if we only find a bicep template, we push the bicep template
- if we find both a bicep & ARM json template, we push the bicep template
- if we only find an ARM JSON template, we push the ARM JSON template

**Comment by AlexanderSehr** [comment]
I'll go ahead an close the issue. @msc365admin if you would not agree, please re-open the issue again. @krbar did test the scenario successfully, so I hope we ironed this out. Thank you very much for your help in raising & supporting us in fixing it.

**Comment by msc365admin** [comment]
Hi @AlexanderSehr, thanks for the feedback!

I do agree with the more refined handling of having both the `main.bicep` & `main.json` fix in #3554, thanks to @krbar. But I don't see a fix for the `404` issue. Am I missing something? I found a duplicate (#3105) of this issue, reported by @msfreaks.

:+1: Kudos for @krbar.
